### ISSUE TITLE
Inverted colors

### ISSUE THREAD

**Issue by Bysmyyr** [swaywm/sway#6577]
I am currently using i3 and using compton to do (smart) color invert with these scripts: https://github.com/vn971/linux-color-inversion I would like to move to Sway but this is a blocker for me. I found similar feature from wayfire and the actual code is quite short( https://github.com/WayfireWM/wayfire/blob/bb8d1ab4703be38a154e769958521901f1116c96/plugins/single_plugins/invert.cpp ) so I hope it would be quite easy thing to do. The feature is awesome, it helps a lot when you need to do long coding sessions and inverting all programs separately is not that easy than just inverting everything. Those smart invert profiles could be very nice but not mandatory.
 
 Any comments? I hope it looks feasible and possible to be added? Ofcourse if this could be done with pluggable script/program, I would be happy to hear that.

**Comment by emersion** [comment]
Maybe the gamma table can be used for this?

Also ref https://github.com/swaywm/sway/issues/3652

**Comment by Bysmyyr** [comment]
gamma table? Could I get a link so I can get myself familiarized.

**Comment by emersion** [comment]
I have a hard time finding references on the topic, but it's a look-up table that can be used to transform colors. Think of it as a function: `output color = gamma_table(input color)`.

The gamma LUT contains 3 curves, each curve is an array of `uint16_t` values. The identity gamma table is built like this: https://github.com/swaywm/wlroots/blob/cdaab820201d6aff7ed44da35595df65b9739bea/backend/drm/legacy.c#L173

I think reversing the identity gamma table could get your inverted colors.

This demo client sets the gamma table: https://github.com/swaywm/wlroots/blob/cdaab820201d6aff7ed44da35595df65b9739bea/examples/gamma-control.c

**Comment by progandy** [comment]
The gamma-control sample should be able to invert the screen with ```gamma=1```, ```contrast=-1``` and ```brightness=2```

That won't allow the "intelligent" hue preserving mode, though.

**Comment by Bysmyyr** [comment]
good to know, thats good start and allows me to use sway daily

**Comment by arbitrary-dev** [comment]
What is gamma-control? Can I `bindsym` it? Or is it about changing some sources?

Color inversion per window would be a cool feature to have in Sway WM.

**Comment by Bysmyyr** [comment]
yes, this needs changing the sources and I had had no time to check this further, feel free.

**Comment by jbeich** [comment]
I use https://github.com/MaxVerevkin/wl-gammarelay-rs/commit/ca76b76632a6 on FreeBSD e.g.,
```
$ pkg install wl-gammarelay-rs basu
$ if [ ! -e $XDG_RUNTIME_DIR/bus ]; then dbus-daemon --session --fork --address=unix:runtime=yes; fi
$ daemon -f wl-gammarelay-rs
$ dbus-send --session --type=method_call --dest=rs.wl-gammarelay / rs.wl.gammarelay.ToggleInverted
$ basuctl --user call rs.wl-gammarelay / rs.wl.gammarelay ToggleInverted
```

**Comment by joh4nd** [comment]
> I use [MaxVerevkin/wl-gammarelay-rs@ca76b76632a6](https://github.com/MaxVerevkin/wl-gammarelay-rs/commit/ca76b76632a6) on FreeBSD e.g.,
> 
> ```shell
> $ doas pkg install wl-gammarelay-rs dbus
> $ if [ -z "$DBUS_SESSION_BUS_ADDRESS" -a ! -e "$XDG_RUNTIME_DIR/bus" ]; then
>     # Prepare session/user bus where wl-gammarelay-rs will advertise its service
>     dbus-daemon --session --fork --address=unix:runtime=yes
>   fi
> $ daemon -f wl-gammarelay-rs
> 
> # freedesktop.org style
> $ dbus-send --session --type=method_call --dest=rs.wl-gammarelay / rs.wl.gammarelay.ToggleInverted
> 
> # GTK style
> $ doas pkg install glib
> $ gdbus call --session --dest=rs.wl-gammarelay --object-path=/ --method=rs.wl.gammarelay.ToggleInverted
> 
> # Qt style
> $ doas pkg install qt5-qdbus qt6-tools
> $ qdbus-qt5 rs.wl-gammarelay / rs.wl.gammarelay.ToggleInverted 
> $ qdbus6 rs.wl-gammarelay / rs.wl.gammarelay.ToggleInverted
> 
> # systemd style
> $ doas pkg install basu
> $ basuctl --user call rs.wl-gammarelay / rs.wl.gammarelay ToggleInverted
> ```

What do you do to install/implement it (I am new with Linux)? I asked for a [tool to invert colors on Wayland](https://unix.stackexchange.com/q/743252/530603), and found this, but haven't tried.

**Comment by nalply** [comment]
`wl-gammactl` (https://github.com/mischw/wl-gammactl) should work. I installed with with `yay -S wl-gammactl`.

However inverting does not work.

```
> wl-gammactl -b 2 -c 1
failed to set gamma table
```

I looked at the source code, made some debugging changes and started gdb. I found out that the options `-b 2 -c 1` creates a linearly descending lookup table, this means it inverts all colors. However it fails at the call `wl_display_roundtrip()`.

I suspect a driver problem. My laptop chokes at a lookup table where values go down. I am not sure. It's a Framework laptop with Iris Xe graphics.

I don't know how to begin. Which project is responsible for the gamma table? swaywm? wlroots? wl-gammactl?

Thanks for any clarification!

**Comment by emersion** [comment]
Yeah, Intel GPUs don't support gamma tables going down. It's a hardware limitation.

A fix would be to implement gamma tables in a shader in wlroots. Also related is ICC profile support, which would allow for the same thing.

**Comment by nalply** [comment]
Thanks for the quick answer! It's a pity.

What happened with wlroots? It is archived, did sway take over wlroots?

**Comment by emersion** [comment]
It moved over to gitlab.freedesktop.org. See the README.

**Comment by emersion** [comment]
Closing because there are multiple ways to achieve this now (either via the protocol or via an ICC profile).

**Comment by CyberShadow** [comment]
@emersion Thank you for the update, this has been one of my blockers for moving from i3. However could you please post more details about how this can be achieved today?

- How can it be achieved using the protocol or with an ICC profile?
- Is it doable selectively per-window?
- Are any other kinds of transformations beyond inverting colors possible, such as the "smart" color inversion described in the original issue with preserves hue but inverts brightness?

**Comment by emersion** [comment]
>    How can it be achieved using the protocol or with an ICC profile?

One would need to generate an ICC profile with a reverse LUT, or apply a reverse LUT via the gamma-control protocol. Unfortunately I haven't found a pre-made ICC profile online.

> Is it doable selectively per-window?

No

> Are any other kinds of transformations beyond inverting colors possible, such as the "smart" color inversion described in the original issue with preserves hue but inverts brightness?

No

**Comment by nalply** [comment]
It worked for me. I had to install vulkan and do the steps described in: https://github.com/swaywm/sway/issues/1486#issuecomment-2344740148, and it works even on the Iris Xe that didn't support gamma tables going down because of some hardware limitation.

THANK YOU!

Note on Arch Linux: do follow all the steps, especially install sway-git! It took some time to get it working for me, it seems to be bleeding edge.

**Comment by nalply** [comment]
Just for clarification, what's the position of wl-gammactl ? I mean, does wl-gammactl change something on Vulkan? I still can't use wl-gammactl to invert colors, it works with the ICC profile, but not with wl-gammactl (but it's fine this way).

**Comment by emersion** [comment]
Right now the wlr-gamma-control protocol and ICC profiles use a different mechanism. wlr-gamma-control sets gamma LUTs in the display hardware, while ICC profiles are applied in a Vulkan shader. Gamma LUTs are much more limited (they can only represent basic color transforms, and as you've noticed some hardware has limitations) but they are more power-efficient.

Ideally we'd add support for applying gamma LUTs in shaders when the display hardware doesn't support it:
https://gitlab.freedesktop.org/wlroots/wlroots/-/merge_requests/4796
